src/scale_impls/decode.rs

```diff
@@ -13,239 +13,296 @@
 /// depending on what was decoded.
-pub fn decode_value_as_type(
+pub fn decode_value_as_type<R: TypeResolver>(
     data: &mut &[u8],
-    ty_id: TypeId,
-    types: &PortableRegistry,
-) -> Result<Value<TypeId>, DecodeError> {
-    scale_decode::visitor::decode_with_visitor(data, ty_id, types, DecodeValueVisitor)
+    ty_id: R::TypeId,
+    types: &R,
+) -> Result<Value<R::TypeId>, DecodeError>
+where
+    R::TypeId: Clone,
+{
+    scale_decode::visitor::decode_with_visitor(
+        data,
+        &ty_id,
+        types,
+        // note: in this case the `FromMapper` converts the u32 into a u32, and is just an identity mapping.
+        DecodeValueVisitor::<R, R::TypeId, FromMapper>::new(),
+    )
 }
 
 // Sequences, Tuples and Arrays all have the same methods, so decode them in the same way:
 macro_rules! to_unnamed_composite {
     ($value:ident, $type_id:ident) => {{
         let mut vals = Vec::with_capacity($value.remaining());
-        while let Some(val) = $value.decode_item(DecodeValueVisitor) {
+        while let Some(val) = $value.decode_item(DecodeValueVisitor::<R, T, F>::new()) {
             let val = val?;
             vals.push(val);
         }
-        Ok(Value { value: ValueDef::Composite(Composite::Unnamed(vals)), context: $type_id.0 })
+        Ok(Value {
+            value: ValueDef::Composite(Composite::Unnamed(vals)),
+            context: F::map($type_id),
+        })
     }};
 }
 
 // We can't implement this on `Value<TypeId>` because we have no TypeId to assign to the value.
-impl scale_decode::DecodeAsFields for Composite<TypeId> {
-    fn decode_as_fields<'info>(
+impl scale_decode::DecodeAsFields for Composite<()> {
+    fn decode_as_fields<'resolver, R: TypeResolver>(
         input: &mut &[u8],
-        fields: &mut dyn FieldIter<'info>,
-        types: &'info PortableRegistry,
+        fields: &mut dyn FieldIter<'resolver, R::TypeId>,
+        types: &'resolver R,
     ) -> Result<Self, scale_decode::Error> {
         // Build a Composite type to pass to a one-off visitor:
-        static EMPTY_PATH: &Path<PortableForm> = &Path { segments: Vec::new() };
         let mut composite =
-            scale_decode::visitor::types::Composite::new(input, EMPTY_PATH, fields, types, false);
+            scale_decode::visitor::types::Composite::new(input, fields, types, false);
         // Decode into a Composite value from this:
-        let val = visit_composite(&mut composite);
+        let val = visit_composite::<R, (), DefaultMapper>(&mut composite);
         // Consume remaining bytes and update input cursor:
         composite.skip_decoding()?;
         *input = composite.bytes_from_undecoded();
-
-        val.map_err(From::from)
+        val.map_err(From::from).map(|v| v.map_context(|_| {}))
     }
 }
 
 /// A [`scale_decode::Visitor`] implementation for decoding into [`Value`]s.
-pub struct DecodeValueVisitor;
+pub struct DecodeValueVisitor<R: TypeResolver, T, F> {
```

Comment:
Instead of having `T` here it'd prob be simpler just to reference `R::TypeId` instead below; one less generic param :)

Likewise we don't use `F` in any meaningful way by the looks of it; can we get rid of all of the `FromMapper` related bits now we are just always handing back `Value<R::TypeId>` or is it still needed?

Reply:
I could not make it work with just `R`, because of this implementation:

```
impl scale_decode::IntoVisitor for Value<()> {
    type AnyVisitor<R: scale_decode::TypeResolver> =
        scale_decode::visitor::VisitorWithCrateError<DecodeValueVisitor<R, (), DefaultMapper>>;

    fn into_visitor<R: scale_decode::TypeResolver>() -> Self::AnyVisitor<R> {
        scale_decode::visitor::VisitorWithCrateError(DecodeValueVisitor::new())
    }
}
```

The issue is, that the `AnyVisitor` type needs to always return `Value<()>` here for any `R: TypeResolver`. If we implement `Visitor` for `DecodeValueVisitor<R>` to always return `Value<R::TypeId>` (nice and reasonable, I agree), the `IntoVisitor` impl does not work anymore.
So I had to add these two sad little type parameters. :(

Reply:
That all makes sense; alas we cannay get rid of the extra params at the mo :)